Servers/FedBeServer.py

```diff
@@ -0,0 +1,121 @@
+from Servers.ABCServer import ABCServer
+import torch
+from torch.distributions import Normal
+from torch.nn import NLLLoss
+from torchcontrib.optim import SWA
+from torch.optim import SGD
+import random
+import numpy as np
+import warnings
+warnings.filterwarnings("ignore")
+
+class FedBeServer(ABCServer):
+
+    def __init__(self,model,loc_data,M=10,swa_lr1=0.001,swa_lr2=0.0004,swa_batch_size=20,
+                    swa_freq=25,seed=1,swa_epochs=1,verbose=True):
+        super().__init__(model)
+        w = model.get_weights()
+        self.loc_data = loc_data
+        # Set initial distributions
+        w_flat = torch.cat([w[k].flatten() for k in w])
+        self.distribution = Normal(w_flat,torch.ones(len(w_flat)))
+        # Set tensorlengths for future reconstruction of flattening.
+        self.tens_lengths = [len(w[k].flatten()) for k in w]
+        self.model_shapes = [w[k].size() for k in w]
+
+        # Set other parameters
+        self.M=M
+        self.swa_lr1=swa_lr1
+        self.swa_lr2=swa_lr2
+        self.swa_batch_size=swa_batch_size
+        self.swa_freq= swa_freq
+        self.swa_epochs=swa_epochs
+        self.verbose = verbose
+        self.seed = seed
+
+
+    def aggregate(self, clients):
+        # Dynamically calculate mean and variance of server weights.
+        if self.verbose: print('FedBe: Reading in clients.')
+        mu_n = 0
+        s_n = 0
+        M_n = 0
+        S=[]
+        for n,client in enumerate(clients):
+            w = client.get_weights()
+            w_flat = torch.cat([w[k].flatten() for k in w])
+            mu_n2 = (n*mu_n+w_flat)/(n+1)
+            M_n= M_n+(w_flat-mu_n)*(w_flat-mu_n2)
+            if(n!=0):
+                s_n= torch.sqrt(M_n/n)
+            mu_n = mu_n2
+            S.append(w)
+
+        # Add small variance to avoid zero variance
+        s_n = s_n+0.0000001
+        self.distribution = Normal(mu_n,s_n)
```

Comment:
Standardize mu towards the amount of data.

Reply:
This is not possible considering current design. Possible add a client.data_size if we want to add this feature.